### ISSUE TITLE
Need help for INT2 quantization

### ISSUE THREAD

**Issue by Thomastw123** [Xilinx/brevitas#1237]
Hello, I am new to Brevitas and having trouble quantized my model to INT2. When I try to do PTQ for the PointPillars LiDAR  model, no classifications are found using INT2 quantization and it returns and empty tuple. So I suspect there is something wrong with quantization.

 Is this the correct way we handle INT2 Quantization?

******I define INT2 quantizer like this*****

class Int2WeightPerTensorFloat(WeightQuantSolver):
    quant_type = QuantType.INT # integer quantization
    bit_width_impl_type = BitWidthImplType.CONST # constant bit width
    float_to_int_impl_type = FloatToIntImplType.ROUND # round to nearest
    scaling_impl_type = ScalingImplType.STATS # scale based on statistics
    scaling_stats_op = StatsOp.MAX # scale statistics is the absmax value
    restrict_scaling_type = RestrictValueType.FP # scale factor is a floating point value
    scaling_per_output_channel = False # scale is per tensor
    bit_width = 2 # bit width is 8
    signed = True # quantization range is signed
    narrow_range = True # quantization range is [-127,127] rather than [-128, 127]
    zero_point_impl = ZeroZeroPoint # zero point is 0.

class Int2ActPerTensorFloat(ActQuantSolver):
    quant_type = QuantType.INT # integer quantization
    bit_width_impl_type = BitWidthImplType.CONST # constant bit width
    float_to_int_impl_type = FloatToIntImplType.ROUND # round to nearest
    scaling_impl_type = ScalingImplType.PARAMETER_FROM_STATS # scale is a parameter initialized from statistics
    scaling_stats_op = StatsOp.PERCENTILE # scale statistics is a percentile of the abs value
    high_percentile_q = 99.999 # percentile is 99.999
    collect_stats_steps = 300  # statistics are collected for 300 forward steps before switching to a learned parameter
    restrict_scaling_type = RestrictValueType.FP # scale is a floating-point value
    scaling_per_output_channel = False  # scale is per tensor
    bit_width = 2  # bit width is 8
    signed = True # quantization range is signed
    narrow_range = False # quantization range is [-128, 127] rather than [-127, 127]
    zero_point_impl = ZeroZeroPoint # zero point is 0.

******Conv2d and Relu*******

 class Backbone(nn.Module):
    def __init__(self, in_channel, out_channels, layer_nums, layer_strides=[2, 2, 2]):
        super().__init__()
        assert len(out_channels) == len(layer_nums)
        assert len(out_channels) == len(layer_strides)
        
        self.multi_blocks = nn.ModuleList()
        for i in range(len(layer_strides)):
            blocks = []
            # Quantized conv block with BatchNorm and activation.
            blocks.append(QuantConv2d(
                in_channels=in_channel, 
                out_channels=out_channels[i], 
                kernel_size=3, 
                stride=layer_strides[i], 
                bias=False, 
                padding=1,
                weight_bit_width=2,
                weight_quant=Int2WeightPerTensorFloat,
                act_quant=Int2ActPerTensorFloat,
                return_quant_tensor=False
            ))
            blocks.append(nn.BatchNorm2d(out_channels[i], eps=1e-3, momentum=0.01))
            blocks.append(QuantReLU(inplace=True, bit_width=2, act_quant=Uint2ActPerTensorFloat, return_quant_tensor=False))

            for _ in range(layer_nums[i]):
                blocks.append(QuantConv2d(
                    in_channels=out_channels[i], 
                    out_channels=out_channels[i], 
                    kernel_size=3, 
                    bias=False, 
                    padding=1,
                    weight_bit_width=2,
                    weight_quant=Int2WeightPerTensorFloat,
                    act_quant=Int2ActPerTensorFloat,
                    return_quant_tensor=False
                ))
                blocks.append(nn.BatchNorm2d(out_channels[i], eps=1e-3, momentum=0.01))
                blocks.append(QuantReLU(inplace=True, bit_width=2, act_quant=Uint2ActPerTensorFloat, return_quant_tensor=False))

            in_channel = out_channels[i]
            self.multi_blocks.append(nn.Sequential(*blocks))

        for m in self.modules():
            if hasattr(m, 'weight') and isinstance(m, QuantConv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')

    def forward(self, x):
        outs = []
        for block in self.multi_blocks:
            x = block(x)
            outs.append(x)
        return outs

**Calibrations**

with torch.no_grad(), calibration_mode(model):
        for i, batched_pts in enumerate(dataloader):
            batch_start = time.time()
            print(f"Processing calibration batch {i+1}/{len(dataloader)}")
            pillars, coors_batch, npoints_per_pillar = model.pillar_layer(batched_pts)
            pillar_features = model.pillar_encoder(pillars, coors_batch, npoints_per_pillar)
            x = model.backbone(pillar_features)
            xs = model.neck(x)
            c = model.head(xs)
            batch_end = time.time()
            print(f"Batch {i+1} processed in {batch_end - batch_start:.2f} seconds")
            if args.max_batches > 0 and (i+1) >= args.max_batches:
                print(f"Stopping calibration early after {args.max_batches} batches.")
                break
    for module in model.modules():
        if hasattr(module, 'quant_act') and hasattr(module.quant_act, 'calculate_amax'):
            module.quant_act.calculate_amax()

**Comment by nickfraser** [comment]
Thanks for the detailed issue. I don't see anything obvious that you have done that's _incorrect_\*, but I have the following observations that you may want to consider:
 - You're using `narrow_range=True` on your weight quantizer, so your model actually has ternary weights, not Int2, e.g., it has 3 unique values, not 4
 - You're using `signed=True` (might be a typo - I see you use a different quantizer in your code) for your activation quantizer, which will effectively be binary for unsigned activations, e.g., ReLU
 - You're using a very high `high_percentile_q = 99.999` value - you're selecting an activation scaling which will _not saturate_ on 99.999% of activation values. If your activations have a Gaussian distribution, then this value may be dictated by outliers, causing many (most?) activations to underflow to 0
 - In general, Int2 quantization is _extremely_ difficult, I would strongly recommend using QAT instead of PTQ. If QAT is not possible for some reason, I would strongly recommend using some more advanced PTQ techniques (e.g., GPTQ, learned round - calibration is really the most basic technique). See our [imagenet examples](https://github.com/Xilinx/brevitas/tree/master/src/brevitas_examples/imagenet_classification/ptq) to help you see how to apply these to your problem. Furthermore, I would strongly recommend starting at a higher precision and gradually reducing it to gauge your progress. For most models, the simple calibration technique you're applying would likely work for ~8-bit models - I've rarely seen it work below 6-bits. What bitwidths can you get to currently? Are you able to reduce this by applying more PTQ techniques or QAT?

\* What you have done may or may not be right, depending on your problem and preferences.

**Comment by nickfraser** [comment]
Closed for inactivity. Please reopen if this is still an issue.
